### PR TITLE
chore(observability): Implement async enterprise configuration reporting

### PR DESCRIPTION
**Pull request by 001wwang**
Closes [OP-382](https://datadoghq.atlassian.net/browse/OP-382)

This PR implements the [async reporting RFC ](https://github.com/DataDog/vector-enterprise/blob/master/rfcs/2022-05-02-OP-331-async-config-reporting.md). In summary, changes include
- Configuration reporting is now handled by a long running task, decoupled from Vector start up/reload
- `exit_on_fatal_error` option removed
- Missing API key prevents start up
- Updating tests

### CODE REVIEW

File: src/app.rs [hunk 1]
```diff
@@ -209,7 +207,7 @@ impl Application {
                         Some(enterprise)
                     }
                     Err(EnterpriseError::MissingApiKey) => {
-                        error!(message = "Enterprise configuration incomplete: missing API key");
+                        error!("Enterprise configuration incomplete: missing API key.");
```
[the remark this answers is not in the document]
Comment (001wwang) [reply src/app.rs:210]: Do you mean use the internal events framework? This occurs during the start up phase where it looks like existing convention is to log directly and return an exit code. I can convert this to an internal event if we'd like.

**Review by leebenson — CHANGES_REQUESTED** [review]
There appears to be at least one condition where a config isn't reported:

If a network isn't available (e.g. wi-fi is explicitly off), the following logs are observed:

```
➜  vector git:(OP-382-api-implement-async-reporting-rfc) target/debug/vector -c ~/Desktop/working.toml
2022-05-24T11:32:14.472352Z  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info,buffers=info,kube=info"
2022-05-24T11:32:14.472984Z  INFO vector::app: Loading configs. paths=["/Users/lee.benson/Desktop/working.toml"]
2022-05-24T11:32:14.484171Z  INFO vector::config::enterprise: Datadog API key provided. Integration with Datadog Observability Pipelines is enabled.
2022-05-24T11:32:14.496923Z  INFO vector::config::enterprise: Attempting to report configuration to Datadog Observability Pipelines.
2022-05-24T11:32:14.501264Z ERROR http: vector::internal_events::http_client: HTTP error. error=error trying to connect: dns error: failed to lookup address information: nodename nor servname provided, or not known error_type="request_failed" stage="processing"
2022-05-24T11:32:14.501689Z ERROR vector::config::enterprise: Could not report Vector config to Datadog Observability Pipelines. err="Failed to make HTTP(S) request: error trying to connect: dns error: failed to lookup address information: nodename nor servname provided, or not known"
2022-05-24T11:32:15.064623Z  INFO vector::topology::running: Running healthchecks.
2022-05-24T11:32:15.066044Z  INFO vector::topology::builder: Healthcheck: Passed.
2022-05-24T11:32:15.066913Z  INFO vector: Vector has started. debug="true" version="0.22.0" arch="x86_64" build_id="none"
2022-05-24T11:32:15.067127Z  INFO vector::app: API is disabled, enable by setting `api.enabled` to `true` and use commands like `vector top`.
2022-05-24T11:32:15.067427Z  INFO vector::sinks::blackhole::sink: Total events collected events=0 raw_bytes_collected=0
2022-05-24T11:32:15.067400Z ERROR http: vector::internal_events::http_client: HTTP error. error=error trying to connect: dns error: failed to lookup address information: nodename nor servname provided, or not known error_type="request_failed" stage="processing"
2022-05-24T11:32:15.067669Z ERROR vector::topology::builder: msg="Healthcheck: Failed Reason." error=Failed to make HTTP(S) request: error trying to connect: dns error: failed to lookup address information: nodename nor servname provided, or not known component_kind="sink" component_type="datadog_metrics" component_id=#datadog_metrics component_name=#datadog_metrics
2022-05-24T11:32:16.068935Z  INFO vector::sinks::blackhole::sink: Total events collected events=1 raw_bytes_collected=514
2022-05-24T11:32:17.068840Z  INFO vector::sinks::blackhole::sink: Total events collected events=1 raw_bytes_collected=514
2022-05-24T11:32:17.092987Z ERROR sink{component_kind="sink" component_id=#datadog_metrics component_type=datadog_metrics component_name=#datadog_metrics}:request{request_id=0}:http: vector::internal_events::http_client: HTTP error. error=error trying to connect: dns error: failed to lookup address information: nodename nor servname provided, or not known error_type="request_failed" stage="processing"
2022-05-24T11:32:17.093623Z  WARN sink{component_kind="sink" component_id=#datadog_metrics component_type=datadog_metrics component_name=#datadog_metrics}:request{request_id=0}: vector::sinks::util::retries: Retrying after error. error=Failed to make HTTP(S) request: error trying to connect: dns error: failed to lookup address information: nodename nor servname provided, or not known
2022-05-24T11:32:18.068421Z  INFO vector::sinks::blackhole::sink: Total events collected events=1 raw_bytes_collected=514
2022-05-24T11:32:18.099382Z ERROR sink{component_kind="sink" component_id=#datadog_metrics component_type=datadog_metrics component_name=#datadog_metrics}:request{request_id=0}:http: vector::internal_events::http_client: HTTP error. error=error trying to connect: dns error: failed to lookup address information: nodename nor servname provided, or not known error_type="request_failed" stage="processing"
```

Of note:

- Config reporting is attempted exactly once.
- Errors/warnings relate to internal 'enterprise' components.
- Turning on the wi-fi adapter sometime later resolves the internal component errors, but the config is never reported.

File: src/app.rs [hunk 2]
```diff
@@ -331,12 +365,27 @@ impl Application {
...
 
                                     #[cfg(feature = "enterprise")]
-                                    // Augment config to enable observability within Datadog, if applicable.
-                                    if let Err(PipelinesError::FatalCouldNotReportConfig) =
-                                        config::enterprise::try_attach(&mut new_config, &config_paths, signal_handler.subscribe()).await
-                                    {
-                                        error!(message = "Shutting down due to configuration reporting failure.");
-                                        break SignalTo::Shutdown;
+                                    match EnterpriseMetadata::try_from(&new_config) {
+                                        Ok(metadata) => {
```
Comment (leebenson) [inline src/app.rs:369]: Could we DRY this up?

**Comment by 001wwang** [comment]
> Of note:
> 
> * Config reporting is attempted exactly once.
> * Errors/warnings relate to internal 'enterprise' components.
> * Turning on the wi-fi adapter sometime later resolves the internal component errors, but the config is never reported.

Thanks for catching this case. Fixed in [43abd24](https://github.com/vectordotdev/vector/pull/12781/commits/43abd24f25a408d9e400c486aab7af1e89d91dab)

File: lib/vector-core/src/config/global_options.rs [hunk 3]
```diff
@@ -39,8 +39,6 @@ pub struct GlobalOptions {
     pub timezone: TimeZone,
     #[serde(skip_serializing_if = "crate::serde::skip_serializing_if_default")]
     pub proxy: ProxyConfig,
-    #[serde(skip)]
-    pub enterprise: bool,
```
Comment (001wwang) [inline lib/vector-core/src/config/global_options.rs:43]: Previously, this was used to control setting DD request headers in the `datadog_logs` sink. However, by coupling enterprise configuration with global options, we prevent reloads that involve adding/removing the `[enterprise]` configuration section. 

The coupling/global option seem unnecessary, so this was refactored into a hidden config option directly on `datadog_logs` sink.

**Review by leebenson — APPROVED** [review]
Tested out the network fix. Works great 🎉

**Pull request merged**